### PR TITLE
build: add clang-tidy configuration file

### PR DESCRIPTION
**Pull request by kylo252**
- add clang-tidy configuration file
- enable additional useful checks through `clangd` that are nice to have (naming-conventions, etc)

references:
   - https://github.com/googleapis/google-cloud-cpp/blob/main/.clang-tidy
   - https://wiki.sei.cmu.edu/confluence/display/c/Clang

---

Related: #10632, #11236, #15428

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -190,6 +190,13 @@ clint: build/.ran-cmake
 clint-full: build/.ran-cmake
 	+$(BUILD_CMD) -C build clint-full
 
+clang-tidy: generate-compilation-db
+	bash scripts/run_clang_tidy.sh $(FILE)
```
Comment (justinmk) [inline Makefile:194]: is the run_clang_tidy.sh script really needed? it's not a pattern we usually follow. Better not to spread things out if possible, also keeps the Makefile more portable (e.g. on Windows)

File: Makefile [hunk 2]
```diff
@@ -190,6 +190,13 @@ clint: build/.ran-cmake
...
+
+generate-compilation-db: build/.ran-cmake
+	ninja -C build -t compdb C_COMPILER__nvim_$(CMAKE_BUILD_TYPE) > build/compile_commands.json	
+	ln -sf build/compile_commands.json .
```
Comment (justinmk) [inline Makefile:198]: looks useful 👀

File: Makefile (hunk 1 above)
Comment (kylo252) [reply Makefile:194]: Only if you want a way to run it per-file from the CLI. If you're running `clangd` then it can also manage `clang-tidy` for you and dynamically display any warnings and code-actions through LSP.

Alternatively, we can introduce another target called  `clang-tidy-file` or something, which would allow us to skip validating the variable inside make.

I didn't want to do conditional run from inside make, because that would get hacky very quickly: https://stackoverflow.com/a/45003119

File: Makefile (hunk 2 above)
Comment (kylo252) [reply Makefile:198]: Yeah, it's the same one I mentioned in #10632

This is essential if we want to run this in the CI and not have it analyze third-party and other non-essential files. 
Let me know if you feel that it's missing some files. You can verify the list quickly with `jq`.

```bash
# you can change the filter or drop it all together
cat compile_commands.json | jq -c '.[] | select( .file | contains("src"))' | wc -l
```

Comment (The-personified-devil) [reply Makefile:198]: Is there a particular reason for creating a symbolic link to `compile_commands.json` in the base directory?
Clangd for example is already configured to search for it in `build/` through `.clangd`.

Comment (kylo252) [reply Makefile:198]: My bad, I didn't notice it. 

It's just the standard way to let the other tools find it without hard-coding the build path. I wanted to use $(BUILD_DIR) but it's not defined.

```make
ln -sf $(BUILD_DIR)/compile_commands.json .
```

There are a few ways for cmake to use it:
1) your typical 
```shell
cd $BUILD_DIR && cmake .. 
```
2) by setting the `CMAKE_BINARY_DIR=..` variable
2) my favorite, but I believe it's relatively new so I need to check which is the oldest supported cmake version
```shell
cmake -B $BUILD_DIR
```

**Review by dundargoc — CHANGES_REQUESTED** [review]
:+1:

File: Makefile (hunk 1 above)
Comment (dundargoc) [reply Makefile:194]: I think this could be done without much hassle inside the makefile. Something like the following should work:

```
FILE ?= $(shell git ls-files *.c)
clang-tidy: generate-compilation-db
	  clang-tidy --config-file .clang-tidy -p build $(FILE)
```

So if it's run as `make clang-tidy` then all files will be used. If it's run with `make clang-tidy FILE=src/nvim/screen.c` then it's only used on `screen.c`.

File: src/nvim/screen.c [hunk 3]
```diff
@@ -1882,7 +1886,9 @@ static int line_putchar(LineState *s, schar_T *dest, int maxcells, bool rl)
   } else {
     if (p_arshape && !p_tbidi && arabic_char(u8c)) {
       // Do Arabic shaping.
-      int pc, pc1, nc;
+      int pc;
+      int pc1;
+      int nc;
```
Comment (justinmk) [inline src/nvim/screen.c:1891]: afaik we decided against this kind of change in the uncrustify discussions.

Comment (gpanders) [reply src/nvim/screen.c:1891]: what about multiple initializations on one line?

```c
int pc = 0, pc1 = 1, nc = 3;
```

IMO these are extremely error prone and ought to be split up. If it's the same rule that governs both then I think we should not do this (but if it's possible to behave differently for initialized and uninitialized declarations then I have no objection).

Comment (justinmk) [reply src/nvim/screen.c:1891]: i'm ok either way. @neovim/core +1 / -1

File: src/nvim/CMakeLists.txt [hunk 4]
```diff
@@ -791,6 +792,22 @@ add_custom_target(
...
+add_custom_target(
+  clint-non-style
+  COMMAND
+    ${LINT_PRG} --filter=-whitespace,-readability ${LINT_NVIM_REL_SOURCES} 
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+  DEPENDS ${LINT_PRG} ${LINT_NVIM_SOURCES} ${LINT_SUPPRESS_FILE}
+)
+
+add_custom_target(
+  clang-tidy-full
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt:804]: why is this needed? please don't add things just because they're marginally nice-to-have. It increases the surface area that developers have to consume.

File: src/nvim/CMakeLists.txt [hunk 5]
```diff
@@ -791,6 +792,22 @@ add_custom_target(
   DEPENDS ${LINT_PRG} ${LINT_NVIM_SOURCES} ${LINT_SUPPRESS_FILE}
 )
 
+add_custom_target(
+  clint-non-style
+  COMMAND
+    ${LINT_PRG} --filter=-whitespace,-readability ${LINT_NVIM_REL_SOURCES}
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt:798]: this is not worth adding as a build target, it is just `clint.py` with some arbitrary options.

File: src/nvim/CMakeLists.txt [hunk 6]
```diff
@@ -63,6 +63,7 @@ set(LINT_SUPPRESS_FILE ${PROJECT_BINARY_DIR}/errors.json)
 set(LINT_SUPPRESS_URL_BASE "https://raw.githubusercontent.com/neovim/doc/gh-pages/reports/clint")
 set(LINT_SUPPRESS_URL "${LINT_SUPPRESS_URL_BASE}/errors.json")
 set(LINT_PRG ${PROJECT_SOURCE_DIR}/src/clint.py)
+set(CLANG_TIDY_CONFIG_FILE ${PROJECT_SOURCE_DIR}/.clang-tidy)
```
Comment (justinmk) [inline src/nvim/CMakeLists.txt:66]: doesn't clang-tidy find this implicitly? that was the reason for putting it in the root dir instead of a subdir like config/

File: Makefile [hunk 7]
```diff
@@ -190,6 +190,18 @@ clint: build/.ran-cmake
 clint-full: build/.ran-cmake
 	+$(BUILD_CMD) -C build clint-full
 
+clang-tidy: generate-compilation-db
+	clang-tidy --config-file .clang-tidy -p build $(FILE)
+
+clang-tidy-full: generate-compilation-db
+	bash -C build clang-tidy-full
```
Comment (justinmk) [inline Makefile:197]: should this be `$(BUILD_CMD)` ?
```suggestion
	+$(BUILD_CMD) -C build clang-tidy-full
```

File: src/nvim/screen.h [hunk 8]
```diff
@@ -59,8 +59,8 @@ extern StlClickDefinition *tab_page_click_defs;
 /// Size of the tab_page_click_defs array
 extern long tab_page_click_defs_size;
 
-#define W_ENDCOL(wp)   (wp->w_wincol + wp->w_width)
-#define W_ENDROW(wp)   (wp->w_winrow + wp->w_height)
+#define W_ENDCOL(wp)   ((wp)->w_wincol + (wp)->w_width)
+#define W_ENDROW(wp)   ((wp)->w_winrow + (wp)->w_height)
```
Comment (justinmk) [inline src/nvim/screen.h:63]: this seems like a good idea

File: src/nvim/screen.c [hunk 9]
```diff
@@ -2116,14 +2122,15 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
...
-#define WL_START        0               // nothing done yet
-# define WL_CMDLINE     WL_START + 1    // cmdline window column
-# define WL_FOLD        WL_CMDLINE + 1  // 'foldcolumn'
-# define WL_SIGN        WL_FOLD + 1     // column for signs
-#define WL_NR           WL_SIGN + 1     // line number
-# define WL_BRI         WL_NR + 1       // 'breakindent'
-# define WL_SBR         WL_BRI + 1       // 'showbreak' or 'diff'
-#define WL_LINE         WL_SBR + 1      // text in the line
+#define WL_START        0                 // nothing done yet
+# define WL_CMDLINE     (WL_START + 1)    // cmdline window column
```
Comment (justinmk) [inline src/nvim/screen.c:2126]: while we are here, the space in `# define` should be removed in all of these cases. It's an accident from the original Vim source import.

File: src/nvim/CMakeLists.txt (hunk 6 above)
Comment (kylo252) [reply src/nvim/CMakeLists.txt:66]: Oh I forgot to point it to the build directory. I also removed the step with linking the database. I'll fix it ASAP.

File: src/nvim/screen.c (hunk 9 above)
Comment (dundargoc) [reply src/nvim/screen.c:2126]: `uncrustify` can fix this. There is a rule that specifically makes the formatting exactly like the style guide (one additional space after # for each level, no space before #).

As for why I haven't fixed it yet...you never asked ;)

File: src/nvim/CMakeLists.txt (hunk 5 above)
Comment (kylo252) [reply src/nvim/CMakeLists.txt:798]: This does not use the `--supress-errors` flag. I thought it could end beng more performant, but I didn't really benchmark it. Anyways, this is totally disposable. I just wanted to make it clear how to get a clearer baseline.

File: src/nvim/screen.c (hunk 3 above)
Comment (kylo252) [reply src/nvim/screen.c:1891]: @gpanders, I don't think the check can make that distinction: https://clang.llvm.org/extra/clang-tidy/checks/readability-isolate-declaration.html

It's quite useful for pointers, so I guess it depends how often do you think it could happen. 

One possibility is to fix the initializer but keep the others, and eventually turn off the check and leave it for the reviewers to enforce the rule.

File: src/nvim/screen.c [hunk 10]
```diff
@@ -7778,20 +7797,20 @@ void limit_screen_size(void)
 
 void win_new_shellsize(void)
 {
-  static long old_Rows = 0;
-  static long old_Columns = 0;
+  static long old_rows = 0;
+  static long old_columns = 0;
```
[the remark this answers is not in the document]
Comment (kylo252) [reply src/nvim/screen.c:7801]: Yeah this is part of the naming convention [rules](https://github.com/neovim/neovim/pull/15601/files#diff-0534891fbc8b89b4c521057e7e8dff70d3cbeb822cc9991a3abad77848c289f8R35-R36): 
> clang-tidy: Invalid case style for variable 'old_Columns' (fix available) [readability-identifier-naming]

File: src/nvim/screen.c (hunk 9 above)
Comment (kylo252) [reply src/nvim/screen.c:2125]: I specifically avoided changing that because I knew it's better to fix it in the uncrustify rules :)

File: src/nvim/screen.c (hunk 3 above)
Comment (gpanders) [reply src/nvim/screen.c:1891]: I'm strongly in favor of simply disallowing multiple declarations on one line.

File: src/nvim/syntax.c [hunk 11]
```diff
@@ -6470,8 +6470,7 @@ const char *const highlight_init_cmdline[] = {
   "default link NvimInvalidOptionSigil NvimInvalidIdentifier",
   "default link NvimInvalidOptionName NvimInvalidIdentifier",
   "default link NvimInvalidOptionScope NvimInvalidIdentifierScope",
-  "default link NvimInvalidOptionScopeDelimiter "
-  "NvimInvalidIdentifierScopeDelimiter",
+  "default link NvimInvalidOptionScopeDelimiter NvimInvalidIdentifierScopeDelimiter",
```
Comment (kylo252) [inline src/nvim/syntax.c:6473]: @justinmk, is there any chance this could be slightly related to https://github.com/neovim/neovim/issues/10241?

Comment (justinmk) [reply src/nvim/syntax.c:6473]: no, but thanks for checking :) This is a good change to make. The line-length is not important enough to justify this quirky formatting.

**Comment by dundargoc** [comment]
Needs rebase. Also not sure if this if this is that useful. But do a rebase and we can check.

File: Makefile [hunk 12]
```diff
@@ -148,6 +148,16 @@ distclean:
...
 	+$(BUILD_TOOL) -C build install
 
+clang-tidy: generate-compilation-db
+	clang-tidy --config-file .clang-tidy -p build $(FILE)
+
+clang-tidy-full: generate-compilation-db
+	+$(BUILD_CMD) -C build clang-tidy-full
+
+generate-compilation-db: build/.ran-cmake
+	ninja -C build -t compdb C_COMPILER__nvim_$(CMAKE_BUILD_TYPE) > build/compile_commands.json
```
Comment (dundargoc) [inline Makefile:158]: Is this needed anymore? Doesn't cmake automatically create a compilation database now?

Comment (kylo252) [reply Makefile:158]: this is specifically for filtering that database, see https://github.com/neovim/neovim/issues/10632#issuecomment-877663416

**Comment by dundargoc** [comment]
Running `make clang-tidy` gives

```
Error: no input files specified.
USAGE: clang-tidy [options] <source0> [... <sourceN>]
```

File: Makefile [hunk 13]
```diff
@@ -148,6 +148,16 @@ distclean:
 install: checkprefix nvim
 	+$(BUILD_TOOL) -C build install
 
+clang-tidy: generate-compilation-db
```
Comment (dundargoc) [inline Makefile:151]: Running `make clang-tidy` gives 

```
Error: no input files specified.
USAGE: clang-tidy [options] <source0> [... <sourceN>]
```

File: Makefile (hunk 12 above)
Comment (dundargoc) [reply Makefile:158]: Wait, that's a completely separate issue. This PR is about introducing clang-tidy, that issue is about "massing" the compilation database for whatever reason.

File: Makefile (hunk 13 above)
Comment (kylo252) [reply Makefile:151]: sorry, this must've broke when I removed the shell script runner. It should be good now.

File: Makefile (hunk 12 above)
Comment (kylo252) [reply Makefile:160]: > for whatever reason

it's specifically for clang gang, see its title :sweat_smile: 
- https://github.com/neovim/neovim/issues/10632

File: .clangd [hunk 14]
```diff
@@ -1,2 +1,32 @@
...
+      - misc-*
+      - performance-*
+      - readability-*
+    Remove:
+      - misc-no-recursion
+      - misc-unused-parameters
+      - readability-function-*
+      - readability-identifier-length
+      - readability-magic-numbers
+      - readability-redundant-function-ptr-dereference
```
Comment (dundargoc) [inline .clangd:18]: I'm not against this per se, but I can imagine it could annoy other developers that this takes priority over their own configurations. We'll need to be cautious not to add too many things at once I think. I'm not saying "remove this", but just want other people's opinions on this.

File: Makefile [hunk 15]
```diff
@@ -148,6 +148,14 @@ distclean:
 install: checkprefix nvim
 	+$(BUILD_TOOL) -C build install
 
+clang-tidy: build/.ran-cmake
+	@test $(FILE) \
+		&& clang-tidy --config-file .clang-tidy -p build $(FILE) \
+		|| $(MAKE) clang-tidy-full
```
Comment (dundargoc) [inline Makefile:154]: What does `test $(FILE)` even do? It gives the same exit code regardless if the file exists or not. Is there a `-f` missing or something?

Also, I think doing 

```suggestion
	clang-tidy --config-file .clang-tidy -p build $(FILE) || $(MAKE) clang-tidy-full
```

would achieve the same thing.

File: Makefile [hunk 16]
```diff
@@ -148,6 +148,14 @@ distclean:
...
+
+clang-tidy-full: build/.ran-cmake
+	+$(BUILD_TOOL) -C build clang-tidy-full
```
Comment (dundargoc) [inline Makefile:157]: You can replace this by adding `clang-tidy-full` to the list with `lintlua lintsh lintuncrustify ...`

Comment (dundargoc) [reply Makefile:157]: Actually, is a make target to run `clang-tidy-full` even needed? There's no way anyone will run this locally since it takes such a long time so I don't think there's much point.

File: .clangd (hunk 14 above)
Comment (dundargoc) [reply .clangd:18]: Actually, maybe we should separate this and move it to a separate PR? It's not really related to clang-tidy so we don't need to think about this at the moment.

Comment (justinmk) [reply .clangd:18]: this is in a ClangTidy section of .clangd. Does it affect LSP results if I'm using clangd? 

In any case, if this matches our guidelines then it makes sense to have the settings defined in our tree.

It looks like some noisy things are being removed here, which is good.

Comment (dundargoc) [reply .clangd:18]: > Does it affect LSP results if I'm using clangd?

Yes.

OK, we can try it out 👍

File: Makefile (hunk 15 above)
Comment (kylo252) [reply Makefile:154]: @dundargoc, did you try it this way?

```sh
make clang-tidy FILE=src/nvim/screen.c
```

File: Makefile (hunk 16 above)
Comment (kylo252) [reply Makefile:157]: I just wanted to follow the steps of clint, but personally I don't think this specific target offers all that much.

File: Makefile (hunk 15 above)
Comment (dundargoc) [reply Makefile]: Yes?

Comment (kylo252) [reply Makefile]: it's good now either way, since [target has been simplified](https://github.com/neovim/neovim/pull/15601/commits/48609e73aacd1c6a104cf97a3836a2561cd8bbe8)

**Review by justinmk — APPROVED** [review]

**Review by lewis6991 — APPROVED** [review]

File: .clangd (hunk 14 above)
Comment (dundargoc) [inline .clangd:18]: I still think the warnings are too noisy. I propose we add the following warnings as well.

```suggestion
      - readability-redundant-function-ptr-dereference
      - readability-isolate-declaration
      - bugprone-assignment-in-if-condition
      - readability-suspicious-call-argument
      - performance-no-int-to-ptr
      - readability-else-after-return
      - google-readability-function-size
      - cert-err33-c
```

**Review by dundargoc — CHANGES_REQUESTED** [review]

File: Makefile [hunk 17]
```diff
@@ -148,6 +148,9 @@ distclean:
 install: checkprefix nvim
 	+$(BUILD_TOOL) -C build install
 
+clang-tidy: generated-sources
+	+clang-tidy --config-file .clang-tidy -p build $(FILE)
```
Comment (lewis6991) [inline Makefile:152]: We should define a default value for `FILE`, even if it's empty.

File: .clangd (hunk 14 above)
Comment (kylo252) [reply .clangd:18]: - `bugprone-assignment-in-if-condition` how often does this even come up and it's not a bug? unless it's some macro
- `cert-err33-c` is pretty serious, see [ERR33-C. Detect and handle standard library errors](https://wiki.sei.cmu.edu/confluence/display/c/ERR33-C.+Detect+and+handle+standard+library+errors)

the rest are style guide related, so I don't mind removing them, although we could only enable them for certain folders/files if they're too noisy for upstream vim files

Comment (dundargoc) [reply .clangd:18]: > how often does this even come up and it's not a bug?

Literally the entire codebase. Did you not test the PR lol?

> cert-err33-c is pretty serious, see [ERR33-C. Detect and handle standard library errors](https://wiki.sei.cmu.edu/confluence/display/c/ERR33-C.+Detect+and+handle+standard+library+errors)

Same thing, it's a valid but it will create massive amounts of noise. This will just result in all of the functions getting a `(void)` slapped on them instead of any error checking.

Comment (kylo252) [reply .clangd:20]: As discussed, this was a a new check that's only available in v15+ which is why I didn't notice it. I've disabled both for now.

File: Makefile (hunk 17 above)
Comment (kylo252) [reply Makefile:152]: After some testing I figured it's not worth it, since clang-tidy will not run in parallel on its own. We can iterate on this further later on.

File: .clangd [hunk 18]
```diff
@@ -1,2 +1,35 @@
...
+    Remove:
+      - google-readability-function-size
+      - misc-no-recursion
+      - misc-unused-parameters
+      - readability-function-*
+      - readability-else-after-return
+      - readability-identifier-length
+      - readability-magic-numbers
+      - readability-redundant-function-ptr-dereference
+      - readability-suspicious-call-argument
```
Comment (dundargoc) [inline .clangd:21]: I've already told you that `readability-isolate-declaration` should be removed. Others have told you in [your other PR](https://github.com/neovim/neovim/pull/20446) that they don't like this change. Why haven't you removed it?

Comment (kylo252) [reply .clangd:21]: oops, sorry! I thought I add it in https://github.com/neovim/neovim/commit/b039593ada338d75aa3f0f80ac66aef6b6656995

File: .clangd [hunk 19]
```diff
@@ -1,2 +1,39 @@
 CompileFlags:
-  CompilationDatabase: build/       # Search build/ directory for compile_commands.json
+  # Search build/ directory for compile_commands.json
+  CompilationDatabase: build
+---
+# Additional useful checks that are too noisy to be included in .clang-tidy
+If:
+  PathMatch: src/nvim/api/.*
```
Comment (dundargoc) [inline .clangd:7]: No, this will make it so that even useful warnings (some of which I rely on) won't show for the regular files since this file takes priority over users own configuration.

This PR is close to being finished, I suggest not making any more changes unless necessary so that we can get this merged. We can iterate on this further in a follow-up PR.

Comment (kylo252) [reply .clangd:7]: just so we're on the same page, this `If` only restricts the following fragments, this means: 
- src/nvim/api: all configured clang-tidy checks +  these defined here (misc, perf, etc)
- everything else: only use configured clang-tidy checks

<details>
<summary>Demo: notice how the "readability-identifier-naming" is only firing for the file in "src/nvim/api"
</summary>

image


</details>

given the above,  you still want me to revert this and make these checks global?

Comment (dundargoc) [reply .clangd:7]: Yes, because there are useful warnings in `.clang-tidy` that should be in `.clangd` instead, which are now ignored. Also, I don't understand why we're separating `src/nvim/api` with the rest of the files, shouldn't they have the same configuration?

Comment (dundargoc) [reply .clangd:7]: Do we actually need a `.clang-tidy`? This will be much easier if we just add only necessary warnings to `.clangd` and then add stuff to `.clang-tidy` in a future PR.

Comment (kylo252) [reply .clangd:7]: > Yes, because there are useful warnings in `.clang-tidy` that should be in `.clangd` instead, which are now ignored.

@dundargoc, this should not be the case, since `.clangd` is not overriding `.clang-tidy` here, just as I explained earlier.

can you please provide an example where that is happening? I've tested on both v14 and v15 and I can't replicate this.

image

> Do we actually need a `.clang-tidy`?

yeah we do since clangd doesn't support all the options AFAIK, see https://github.com/clangd/clangd/issues?q=tidy+is%3Aopen

Comment (dundargoc) [reply .clangd:7]: `src/nvim/memline.c` is a good example: there are a bunch of unnecessary casts there and this PR removes those warnings.

Comment (kylo252) [reply .clangd:7]: oh, okay I think I can see that you mean `readability-non-const-parameter` is useful, where you'd still want it on for all the files.
either way, I'll remove the guard since we can always put it back if it's too noisy.

Comment (ii14) [reply .clangd]: > yeah we do since clangd doesn't support all the options AFAIK, see https://github.com/clangd/clangd/issues?q=tidy+is%3Aopen

@kylo252 what am I supposed to look at here?
